orc/orcarm.c

```diff
@@ -1016,3 +1016,201 @@ orc_arm64_emit_am (OrcCompiler *p, OrcArm64RegBits bits, OrcArm64DP opcode,
+        alias = 0;
+      } else if (imms < immr) {
+        sprintf (opt_immr, ", #%u", immr);
+        sprintf (opt_imms, ", #%u", imms);
+        alias = 1;
+      } else if (bfx_preferred(bits, 0, imms, immr)) {
+        sprintf (opt_immr, ", #%u", immr);
+        sprintf (opt_imms, ", #%u", imms);
+        alias = 2;
+      } else if (immr == 0) {
+        if (imms == 0x7)
+          alias = 3;
+        else if (imms == 0xf)
+          alias = 4;
+        else if (imms == 0x1f)
+          alias = 5;
+      }
+      break;
+    case ORC_ARM64_DP_BFM:
+      if (imms < immr) {
+        if (Rn == 0x1f) {
+          sprintf (opt_immr, ", #%u", immr);
+          alias = 0;
+        } else {
+          sprintf (opt_immr, ", #%u", immr);
+          sprintf (opt_imms, ", #%u", imms);
+          alias = 1;
+        }
+      } else {
+        sprintf (opt_immr, ", #%u", immr);
+        sprintf (opt_imms, ", #%u", imms);
+        alias = 2;
+      }
+      break;
+    case ORC_ARM64_DP_UBFM:
+      if (imms == 0x1f) {
+        if (bits == ORC_ARM64_REG_64)
+          imms |= 0x20;
+        sprintf (opt_immr, ", #%u", immr);
+        alias = 1;
+      } else if (imms + 1 == immr) {
+        if (bits == ORC_ARM64_REG_64)
+          imms |= 0x20;
+        sprintf (opt_immr, ", #%u", immr);
+        alias = 0;
+      } else if (imms < immr) {
+        sprintf (opt_immr, ", #%u", immr);
+        sprintf (opt_imms, ", #%u", imms);
+        alias = 2;
+      } else if (bfx_preferred(bits, 1, imms, immr)) {
+        sprintf (opt_immr, ", #%u", immr);
+        sprintf (opt_imms, ", #%u", imms);
+        alias = 3;
+      } else if (immr == 0) {
+        if (imms == 0x7)
+          alias = 4;
+        else if (imms == 0xf)
+          alias = 5;
+      }
+      break;
+    default:
+      ORC_COMPILER_ERROR(p, "unknown opcode %d", opcode);
+      return;
+  }
+
+  opcode -= ORC_ARM64_DP_SBFM;
+
+  code = arm64_code_bfm(bits, opcode, immr, imms, Rn, Rd);
+
+  if (alias != -1) {
+    ORC_ASM_CODE(p, "  %s %s, %s%s%s\n",
+        insn_alias[opcode][alias],
+        orc_arm64_reg_name(Rd, bits),
+        orc_arm64_reg_name(Rn, bits),
+        opt_immr, opt_imms);
+  } else {
+    ORC_ASM_CODE(p, "  %s %s, %s, #%u, #%u\n",
+        insn_names[opcode],
+        orc_arm64_reg_name(Rd, bits),
+        orc_arm64_reg_name(Rn, bits),
+        immr, imms);
+  }
+
+  orc_arm_emit (p, code);
+}
+
+/** data processing instructions: Extract, alias of ROR (imm)
+ *
+ *    3                   2                   1
+ *  1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |b|0 0|1 0 0 1 1 1|N|0|    Rm   |    imms   |    Rn   |    Rd   |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ */
+
+#define arm64_code_extr(b,Rm,imms,Rn,Rd) (0x13800000 | (((b==64)&0x1)<<31) | \
+    (((b==64)&0x1)<<22) | (((Rm)&0x1f)<<16) | (((imms)&0x3f)<<10) | \
```

Comment:
wrapping b with () is also recommended here.

Reply:
Updated